crates/wasmparser/src/validator/component.rs

```diff
@@ -0,0 +1,1800 @@
+                            ComponentEntityType::from_type_def(
+                                types.get(i.ty, offset)?.clone(),
+                                "import",
+                                offset,
+                            )?,
+                        )
+                        .is_some()
+                    {
+                        return Err(BinaryReaderError::new(
+                            format!("duplicate import name `{}` already defined", i.name),
+                            offset,
+                        ));
+                    }
+                }
+                crate::ComponentType::OuterType { count, index } => {
+                    let ty = if *count == 0 {
+                        types.get(*index, offset)?.clone()
+                    } else {
+                        let count = *count as usize;
+                        if count > parents.len() {
+                            return Err(BinaryReaderError::new(
+                                format!("invalid outer alias count of {}", count),
+                                offset,
+                            ));
+                        }
+
+                        parents[parents.len() - count]
+                            .types
+                            .get(*index, offset)?
+                            .clone()
+                    };
+
+                    types.0.push(ty);
+                }
+            };
+        }
+
+        Ok(ComponentType {
+            imports: Rc::new(imports),
+            exports: Rc::new(exports),
+        })
+    }
+
+    fn instance_type(
+        defs: &[crate::InstanceType],
+        features: &WasmFeatures,
+        parents: &[ComponentState],
+        offset: usize,
+    ) -> Result<InstanceType> {
+        let mut types = ComponentTypeSpace::default();
+        let mut exports = HashMap::new();
+
+        for def in defs {
+            match def {
+                crate::InstanceType::Type(ty) => {
+                    types.type_def(ty.clone(), features, parents, offset)?;
+                }
+                crate::InstanceType::Export { name, ty } => {
+                    if exports
+                        .insert(
+                            name.to_string(),
+                            ComponentEntityType::from_type_def(
+                                types.get(*ty, offset)?.clone(),
+                                "export",
+                                offset,
+                            )?,
+                        )
+                        .is_some()
+                    {
+                        return Err(BinaryReaderError::new(
+                            format!("duplicate export name `{}` already defined", name),
+                            offset,
+                        ));
+                    }
+                }
+                crate::InstanceType::OuterType { count, index } => {
+                    let ty = if *count == 0 {
```

Comment:
I feel like this may be the 3rd or 4th time I've seen a few things here, so could some helper methods/structs/etc be added to reduce some of the duplciation? Everything around alias lookup seems to be largely open-coded (e.g. the "if 0 self else parent" logic). Additionally for example `component_type` above is very similar to this `instance_type` method and I think it'd be good to reduce some of the duplication there.

There's also a fair number of places along the lines of "add this string to this set and do something on duplicate" which seems like they could be deduplicated or shared somehow?

Reply:
This has been dedup'd with the `check_alias_count` function.